File: Prototype/Constants/ValidationConstants.cs
namespace Prototype.Constants
{
    public static class ValidationConstants
    {
        public const int MinPasswordLength = 8;
        public const int MaxPasswordLength = 50;
        public const int MinUsernameLength = 3;
        public const int MaxUsernameLength = 50;
        public const int MaxEmailLength = 100;
        
        public static class RegexPatterns
        {
            public const string Email = @"^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$";
            public const string Username = @"^[a-zA-Z0-9_-]+$";
            public const string StrongPassword = @"^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]";
            public const string PhoneNumber = @"^\+?[1-9]\d{1,14}$";
        }
        
        public static class ErrorMessages
        {
            public const string RequiredField = "{0} is required";
            public const string InvalidEmail = "Invalid email format";
            public const string InvalidUsername = "Username can only contain letters, numbers, underscores, and hyphens";
            public const string WeakPassword = "Password must contain at least one uppercase letter, one lowercase letter, one number, and one special character";
            public const string PasswordTooShort = "Password must be at least 8 characters long";
            public const string PasswordTooLong = "Password must not exceed 50 characters";
            public const string UsernameTooShort = "Username must be at least 3 characters long";
            public const string UsernameTooLong = "Username must not exceed 50 characters";
        }
    }
}

File: Prototype/ClientApp/src/utils/validationConstants.ts
/**
 * Shared validation constants between frontend and backend
 * These should match the backend ValidationConstants.cs
 */

export const VALIDATION_CONSTANTS = {
  // Password validation
  PASSWORD: {
    MIN_LENGTH: 8,
    MAX_LENGTH: 128, // Standardized to match backend max
    REQUIRE_UPPERCASE: true,
    REQUIRE_LOWERCASE: true,
    REQUIRE_DIGIT: true,
    REQUIRE_SPECIAL_CHAR: false, // Frontend doesn't enforce special chars by default
  },

  // Username validation  
  USERNAME: {
    MIN_LENGTH: 3,
    MAX_LENGTH: 50, // Standardized to match backend ValidationConstants
    PATTERN: /^[a-zA-Z0-9._-]+$/, // Alphanumeric with dots, underscores, hyphens
  },

  // Email validation
  EMAIL: {
    MAX_LENGTH: 100,
    PATTERN: /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/, // Match backend pattern
  },

  // Phone number validation
  PHONE: {
    MIN_DIGITS: 10,
    MAX_DIGITS: 15, // Standardized to be more permissive
    MAX_LENGTH: 20, // With formatting characters
  },

  // Application validation
  APPLICATION: {
    NAME: {
      MIN_LENGTH: 3,
      MAX_LENGTH: 100,
    },
    DESCRIPTION: {
      MIN_LENGTH: 10,
      MAX_LENGTH: 500,
    },
  },

  // General field validation
  FIELD: {
    FIRST_NAME: {
      MIN_LENGTH: 2,
      MAX_LENGTH: 50,
    },
    LAST_NAME: {
      MIN_LENGTH: 2,
      MAX_LENGTH: 50,
    },
  },
} as const;

// Error messages matching backend patterns
export const VALIDATION_MESSAGES = {
  REQUIRED: (fieldName: string) => `${fieldName} is required`,
  MIN_LENGTH: (fieldName: string, min: number) => `${fieldName} must be at least ${min} characters long`,
  MAX_LENGTH: (fieldName: string, max: number) => `${fieldName} must be no more than ${max} characters long`,
  INVALID_EMAIL: 'Please enter a valid email address',
  INVALID_PHONE: 'Please enter a valid phone number',
  PASSWORD_MISMATCH: 'Passwords do not match',
  INVALID_USERNAME: 'Username can only contain letters, numbers, dots, underscores, and hyphens',
  WEAK_PASSWORD: 'Password must contain at least one uppercase letter, one lowercase letter, and one digit',
} as const;